src/AssumeRoleProvider.ts

```diff
@@ -0,0 +1,262 @@
+
+    /**
+     * Internal Tracking variables
+     */
+    this._credentials = null
+  }
+
+  getRequestConfig(): {
+    requestOptions: http.RequestOptions
+    requestData: string
+  } {
+    const hostValue = this.stsEndpoint.hostname
+    const portValue = this.stsEndpoint.port
+    const qryParams = new URLSearchParams({ Action: this.action, Version: '2011-06-15' })
+
+    qryParams.set('DurationSeconds', this.expirySeconds.toString())
+
+    if (this.policy) {
+      qryParams.set('Policy', this.policy)
+    }
+    if (this.roleArn) {
+      qryParams.set('RoleArn', this.roleArn)
+    }
+
+    if (this.roleSessionName != null) {
+      qryParams.set('RoleSessionName', this.roleSessionName)
+    }
+    if (this.token != null) {
+      qryParams.set('Token', this.token)
+    }
+
+    if (this.webIdentityToken) {
+      qryParams.set('WebIdentityToken', this.webIdentityToken)
+    }
+
+    if (this.externalId) {
+      qryParams.set('ExternalId', this.externalId)
+    }
+
+    const urlParams = qryParams.toString()
+    const contentSha256 = toSha256(urlParams)
+
+    const date = new Date()
+
+    const requestOptions = {
+      hostname: hostValue,
+      port: portValue,
+      path: '/',
+      protocol: this.stsEndpoint.protocol,
+      method: 'POST',
+      headers: {
+        'Content-Type': 'application/x-www-form-urlencoded',
+        'content-length': urlParams.length.toString(),
+        host: hostValue,
+        'x-amz-date': makeDateLong(date),
+        'x-amz-content-sha256': contentSha256,
+      } as Record<string, string>,
+      agent: this.transportAgent,
+    } satisfies http.RequestOptions
+
+    requestOptions.headers.authorization = signV4ByServiceName(
+      requestOptions,
+      this.accessKey,
+      this.secretKey,
+      this.region,
+      date,
+      contentSha256,
+      'sts',
+    )
+
+    return {
+      requestOptions,
+      requestData: urlParams,
+    }
+  }
+
+  async performRequest(): Promise<CredentialResponse> {
+    const { requestOptions, requestData } = this.getRequestConfig()
+
+    const res = await request(this.transport, requestOptions, requestData)
+
+    const body = await readAsString(res)
+
+    return parseXml(body)
+  }
+
+  parseCredentials(respObj: CredentialResponse): Credentials {
+    if (respObj.ErrorResponse) {
+      throw new Error(
+        `Unable to obtain credentials: ${respObj.ErrorResponse?.Error?.Code} ${respObj.ErrorResponse?.Error?.Message}`,
+        { cause: respObj },
+      )
+    }
+
+    const {
+      AssumeRoleResponse: {
+        AssumeRoleResult: {
+          Credentials: {
+            AccessKeyId: accessKey,
+            SecretAccessKey: secretKey,
+            SessionToken: sessionToken,
+            Expiration: expiresAt,
+          },
+        },
+      },
+    } = respObj
+
+    this.accessExpiresAt = expiresAt
+
+    return new Credentials({ accessKey, secretKey, sessionToken })
+  }
+
+  async refreshCredentials(): Promise<Credentials> {
```

Comment:
`Client` will throw error if this method return `null`, so just typed as `Promise<Credentials>`